server/test/unit/plugins/file/test_distributor.py

```diff
@@ -345,3 +345,83 @@ def test__symlink_units_EINVAL_os_error(self, mock_readlink):
         self.assertTrue(os.path.islink(original_link))
         created_link = readlink(original_link)
         self.assertNotEqual(old_target, created_link)
+
+    @patch('os.path.join')
+    @patch('os.path.isabs')
+    @patch('os.path.normpath')
+    def test__target_symlink_path_OK_path(self, mock_normpath, mock_isabs, mock_join):
+        """
+        Check an OK target path passes the symlink path validation
+        """
+        distributor = self.create_distributor_with_mocked_api_calls()
+        build_dir = '/foo/bar'
+        target_path = 'baz'
```

Comment:
Suggestion: Should we also test a valid relative path with `..` and `.`, e.g. `foo/.././..baz/.bar`? (Does not need to be such a detailed test like here, just verifying that the expected path is returned by `_target_symlink_path()`

Reply:
I think the `os.path.normpath()` used in `target_symlink_path()` removes the "middle (double) dots" of the path, moving them in front (reducing other directories on the way). A testcase like that would therefore check that `os.path.normpath()` works as expected. The other tests rather assert particular code path in `_target_symlink_path()` were exercised. I can add that tho.

Reply:
> A testcase like that would therefore check that os.path.normpath() works as expected.

Just to elaborate where I am coming from:

Yes. But more precisely, it would check what you or I _expect_ `os.path.normpath()` (and the remaining logic and functions) to do. Perhaps it does not do what we expect. (The how 'gears are wheeling' test does not verify those assumptions because the assumptions are encoded into the test.)

But more importantly, if someone wants to refactor or change the function in the future, it may not be based on `os.path.normpath()` anymore. The tests below describe when the function is expected to raise an exception. The proposed test for the positive case(s) would describe what the function is supposed to do.

In conclusion, I think both tests have their merit, the "inner gears working" one and the "outer expectation" one.